### DISCUSSION TITLE
Klarna integration with Saleor

### DISCUSSION THREAD

**Discussion by satvikt** [saleor/saleor#9165] [General]
For our e-commerce website, We are integrating Saleor with Klarna Checkout - https://docs.klarna.com/klarna-checkout/

**The general flow of the Klarna checkout is:**

1. Klarna checkout replaces the underlying traditional checkout experience
2. The user interacts with Klarna checkout snippet (rendered in an iframe by the storefront), providing the billing address, shipping address and shipping method to Klarna, instead of Saleor.
3. After the user presses the "Pay" button and Klarna processes the payment, Klarna redirects the current page (/checkout) to the confirmation page of the merchant (our) choice.

**Sample checkout page from our React-storefront, with the red-bordered box highlighting the Klarna Checkout Snippet (Rendered into an iframe on the checkout page)**
![klarna_sample](https://user-images.githubusercontent.com/21122115/153910974-eafa9c93-e9c1-472d-9a17-50232ece62d9.png)

**Steps involved in integrating Klarna Checkout**
https://docs.klarna.com/klarna-checkout/

1. Hit Klarna Create Order API with the checkout data to create a Klarna order.
2. Klarna responds with the checkout HTML snippet to be rendered on the front-end.
3. This is provided to the front-end by the server
4. User inputs the billing address, shipping address and shipping method. User selects the desired payment method and clicks on the Pay button,
5. Klarna redirects the front-end to the "confirmation" page - this is supplied as part of the "create order API" payload in step 1, providing the klarna_order_id as a URL param
6. the server needs to call the Klarna Read Order API with klarna_order_id
7. Klarna responds with a "confirmation HTML snippet" which is finally rendered on the "confirmation" page by the front-end

## Our Approach

**We are implementing Klarna as a payment gateway plugin in Saleor same as Adyen or Stripe.** 

> Please note that for the purpose of the below diagram, **core** refers to the code implemented in the Klarna plugin we've created in Saleor core.

**Here's an activity diagram of the designed solution:**

![klarma_checkout_activity_diagram](https://user-images.githubusercontent.com/21122115/153911417-a716ba42-af3d-43f8-93b2-a558d938c4be.png)

### Please note:

1. The ONLY way to get the address/shipping method/payment data from Klarna is via|
- the server-side callbacks which Klarna sends at various point of user interaction
OR
- hitting the Klarna API.

Hence, we had to rule out using mutations like _CheckoutBillingAddressUpdate_, _CheckoutShippingAddressUpdate_, _CheckoutShippingMethodUpdate_ and _CheckoutPaymentCreate_

### Current roadblock:
We need to update the shipping method before the call to CheckoutComplete
_Note - We can't update the shipping method in step 12 as that Klarna callback gives a window of a mere 3 sec before it cancels the order._

### Our questions

1. Is this sample architecture/design in sync with a recommended way for implementing a new payment gateway (Given the difference in Klarna Checkout)
2. Which mutation/plugin hook to use for step 16?
3. If there is no out-of-the-box mutation recommended for this scenario, should we add a custom mutation?
4. If yes, what would be the most non-intrusive way to do this (custom mutation)? 
Please note that **we do not want to make changes in the core so that we have a clean upgrade path available**

**Comment by timuric** [comment]
We are currently working on new mutations to make it simple to facilitate such flows using synchronous webhooks without the need to modify the core or create plugins. If you are interested in the preview of this new API please DM on twitter.

**Reply by satvikt** [reply]
Hi Timur, I have sent you a DM on Twitter. Very curious to see the API preview!

Also, is there any timeline as to when the API will become publicly available?

**Comment by prometx11** [comment]
I will also work on a Klarna integration. Your workflows will help a lot.
If I have some more input for a concrete implementation I will also share it here. 
Maybe we could add the Klarna integration as a plugin for others to use.

**Reply by Bigosss** [reply]
Great idea ! seeing recent popularity of Klarna it would be so nice to have it. Are you working on it?

**Reply by tracure1337** [reply]
This would really be very welcome & helpful.

**Reply by prometixX** [reply]
We didnt work on the intergation since there is already a very good integration with custom apps. 
Have a look at Saleors React-Storefront monorepo with the checkout app: https://github.com/saleor/react-storefront

**Comment by timuric** [comment]
Klarna is available via payment gateways such as mollie and adyen, which are implemented in the storefront demo:
https://github.com/saleor/react-storefront
